game-app/game-core/src/main/java/games/strategy/triplea/ui/TechnologyPanel.java

```diff
@@ -0,0 +1,288 @@
+    final TableColumn techNameColumn = techTable.getColumnModel().getColumn(0);
+    techNameColumn.setHeaderRenderer(techNameComponentRenderer);
+    techNameColumn.setCellRenderer(techNameComponentRenderer);
+    return techTable;
+  }
+
+  static final class TechNameComponentRenderer extends DefaultTableCellRenderer {
+    @Override
+    public void setValue(Object aValue) {
+      setToolTipText(aValue.toString());
+      super.setValue(aValue);
+    }
+  }
+
+  public void setGameData(final GameData data) {
+    gameData.removeDataChangeListener(this);
+    gameData = data;
+    gameData.addDataChangeListener(this);
+    gameDataChanged(null);
+  }
+
+  @Override
+  public void gameDataChanged(final Change change) {
+    techModel.markDirty();
+    SwingUtilities.invokeLater(this::repaint);
+  }
+
+  /**
+   * Gets the small flag for a given PlayerId.
+   *
+   * @param player the player to get the flag for
+   * @return ImageIcon small flag
+   */
+  protected ImageIcon getIcon(final GamePlayer player) {
+    ImageIcon icon = mapPlayerImage.get(player);
+    if (icon == null && uiContext != null) {
+      final Image img = uiContext.getFlagImageFactory().getSmallFlag(player);
+      icon = new ImageIcon(img);
+      icon.setDescription(player.getName());
+      mapPlayerImage.put(player, icon);
+    }
+    return icon;
+  }
+
+  protected @Nullable ImageIcon getIcon(final String playerName) {
+    final GamePlayer player = gameData.getPlayerList().getPlayerId(playerName);
+    if (player == null) {
+      return null;
+    }
+    return getIcon(player);
+  }
+
+  private void fillPlayerIcons() {
+    for (final GamePlayer p : gameData.getPlayerList().getPlayers()) {
+      getIcon(p);
+    }
+  }
+
+  static class JComponentTableCellRenderer implements TableCellRenderer {
+    @Override
+    public Component getTableCellRendererComponent(
+        final JTable table,
+        final Object value,
+        final boolean isSelected,
+        final boolean hasFocus,
+        final int row,
+        final int column) {
+      return (JComponent) value;
+    }
+  }
+
+  class TechTableModel extends AbstractTableModel {
+    private static final long serialVersionUID = -4612476336419396081L;
+    /* Flag to indicate whether data needs to be recalculated */
+    private boolean isDirty = true;
+    /* Column Header Names */
+    /* Row Header Names */
+    private String[] colList;
+    /* Underlying data for the table */
+    private final String[][] data;
+    /* Convenience mapping of country names -> col */
+    private final Map<String, Integer> colMap = new HashMap<>();
+    /* Convenience mapping of technology names -> row */
+    private final Map<String, Integer> rowMap = new HashMap<>();
+
+    TechTableModel() {
+      initColList();
+      /* Load the country -> col mapping */
+      for (int i = 0; i < colList.length; i++) {
+        colMap.put(colList[i], i + 1);
```

Comment:
Maybe add a variable `columnNr = columnId + 1` which explains better why there is a `+ 1` and what is stored in `colMap `here.